flight/PiOS/Common/pios_ms5xxx.c

```diff
@@ -0,0 +1,700 @@
+#include "pios.h"
+
+#if defined(PIOS_INCLUDE_MS5XXX) || defined(PIOS_INCLUDE_MS5XXX_SPI)
+
+#include "pios_ms5xxx_priv.h"
+#include "pios_semaphore.h"
+#include "pios_thread.h"
+#include "pios_queue.h"
+
+#include "physical_constants.h"
+
+/* Private constants */
+#define MS5XXX_TASK_PRIORITY	PIOS_THREAD_PRIO_HIGHEST
+#define MS5XXX_TASK_STACK_BYTES	512
+
+/* Private Variables */
+
+/* Private methods */
+static int32_t PIOS_MS5XXX_Read(uint8_t address, uint8_t * buffer, uint8_t len);
+static int32_t PIOS_MS5XXX_WriteCommand(uint8_t command);
+static void PIOS_MS5XXX_Task(void *parameters);
+
+/* Private types */
+//! The valid hardware buses
+enum PIOS_BUS_TYPE {
+	PIOS_SPI_BUS,
+	PIOS_I2C_BUS,
+};
+
+/* Local Types */
+
+enum PIOS_MS5XXX_DEV_MAGIC {
+	PIOS_MS5XXX_DEV_MAGIC = 0x1c50bcf2, // md5 of `PIOS_MS5XXX_DEV_MAGIC`
+};
+
+enum conversion_type {
+	PRESSURE_CONV,
+	TEMPERATURE_CONV
+};
+
+struct ms5xxx_dev {
+	const struct pios_ms5xxx_cfg * cfg;
+	enum PIOS_BUS_TYPE pios_bus_type; // Bus type can be either SPI or I2C
+	uint32_t dev_id; // Valid for both SPI and I2C
+	uint32_t slave_num; // Only used for SPI bus chip select
+	uint8_t ms5xxx_i2c_addr; // Only used for I2C 7-bit address
+	struct pios_thread *task;
+	struct pios_queue *queue;
+
+	int64_t pressure;
+	int64_t temperature_unscaled;
+	uint16_t calibration[6];
+	enum conversion_type current_conversion_type;
+	enum PIOS_MS5XXX_DEV_MAGIC magic;
+
+	struct pios_semaphore *busy;
+};
+
+static struct ms5xxx_dev *dev;
+
+/**
+ * @brief Allocate a new device
+ */
+static struct ms5xxx_dev * PIOS_MS5XXX_alloc(void)
+{
+	struct ms5xxx_dev *ms5xxx_dev;
+
+	ms5xxx_dev = (struct ms5xxx_dev *)PIOS_malloc(sizeof(*ms5xxx_dev));
+	if (!ms5xxx_dev) {
+		return (NULL);
+	}
+
+	memset(ms5xxx_dev, 0, sizeof(*ms5xxx_dev));
+
+	ms5xxx_dev->queue = PIOS_Queue_Create(1, sizeof(struct pios_sensor_baro_data));
+	if (ms5xxx_dev->queue == NULL) {
+		PIOS_free(ms5xxx_dev);
+		return NULL;
+	}
+
+	ms5xxx_dev->magic = PIOS_MS5XXX_DEV_MAGIC;
+
+	ms5xxx_dev->busy = PIOS_Semaphore_Create();
+	PIOS_Assert(ms5xxx_dev->busy != NULL);
+
+	return ms5xxx_dev;
+}
+
+/**
+ * @brief Validate the handle to the device
+ * @returns 0 for valid device or <0 otherwise
+ */
+static int32_t PIOS_MS5XXX_Validate(struct ms5xxx_dev *dev)
+{
+	if (dev == NULL)
+		return -1;
+	if (dev->magic != PIOS_MS5XXX_DEV_MAGIC)
+		return -2;
+	if (dev->dev_id == 0)
+		return -3;
+	return 0;
+}
+
+
+/**
+ * @brief PIOS_MS5XXX_Init Initializes the MS5xxx chip
+ * @param cfg Configuration structure. Note that it is not deep copied.
+ * @return 0 on success. -1 if the allocation fails, -2 if the board
+ * setup file does not define the model of chip, and -3 if the write
+ * command fails.
+ */
+int32_t PIOS_MS5XXX_Init(const struct pios_ms5xxx_cfg *cfg)
+{
+	// Set the pointer to the configuration struct. Note that it is not deep copied.
+	dev->cfg = cfg;
+
+	// Check that a model of the chip was chosen
+	if (cfg->pios_ms5xxx_model == 0) {
```

Comment:
Alas, not. I didn't see a WHOAMI or something that would be unique to each different model in the series. That's a real shame, because you will get the wrong absolute altitude if you use the wrong driver math, but your relative altitude changes will still be sort-of correct. (Ask me how I know!)